File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def helper(self,r,c,grid,dp):
        if(grid[r][c]==1): return 0
        if(r==len(grid)-1 and c==len(grid[0])-1):
            return 1
        
        if dp[r][c]!=-1:
            return dp[r][c]

        path=0
        if(r<len(grid)-1):
            path+=self.helper(r+1,c,grid,dp)
        if(c<len(grid[0])-1):
            path+=self.helper(r,c+1,grid,dp)
        dp[r][c]=path
        return path

    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        if obstacleGrid[0][0] == 1:
            return 0

        dp=[[-1 for _ in range(len(obstacleGrid[0]))] for _ in range(len(obstacleGrid))]   
        return self.helper(0,0,obstacleGrid,dp)
    
        

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        int r=grid.size();
        int c=grid[0].size();
        vector<vector<int>> dp(r,vector<int>(c,0));
    
        if(grid[0][0]==1)
            return 0;

        dp[0][0]=1;
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                if(grid[i][j]==1){
                    dp[i][j]=0;
                    continue;
                }

                if(i>0)
                    dp[i][j]=dp[i][j]+dp[i-1][j];
                if(j>0)
                    dp[i][j]=dp[i][j]+dp[i][j-1];
            }
        }
        return dp[r-1][c-1];
    }
};